lib/modules/manager/cargo/artifacts.ts

```diff
@@ -64,11 +85,27 @@ export async function updateArtifacts({
   try {
     await writeLocalFile(packageFileName, newPackageFileContent);
     logger.debug('Updating ' + lockFileName);
-    await cargoUpdate(
-      packageFileName,
-      isLockFileMaintenance,
-      config.constraints?.rust
-    );
+
+    if (isLockFileMaintenance) {
+      await cargoUpdate(packageFileName, true, config.constraints?.rust);
+    } else if (updatedDeps.find(dep => dep.lockedVersion)) {
+      // If we can identify packages by their locked version, then update them precisely
+      for (const dep of updatedDeps) {
```

Comment:
move loop inside the function and pass all commands to a single exec call, you can pass a string array (perf)

otherwise it would cause multiple docker calls and multiple rust installs on each exec on docker binary source.

Reply:
Unfortunately you can't batch cargo updates if `--precise` is specified. I guess we could batch it ourselves through `sh`/`bash` though.

Reply:
our exec already supports batches, just pass an array with all required commands and renovate will do it.